### DISCUSSION TITLE
Connection Flow clarification

### DISCUSSION THREAD

**Discussion by carillem** [Reedyuk/blue-falcon#139] [Q&A] [answered]
As far as I understand, this is the flow but I am unsure if `readCharacteristic`, `notifyCharacteristic`, `writeCharacteristic`, and `readDescriptor` are setter functions and waht order I'm supposed to call them to complete my setup?

- scan()
- stopScanning()
- connect()
 - readRemoteRssi()
 - discoverServices()
 - discoverCharacteristics() // At this point I am able to see the services, characteristics, and see the descriptor
 - // TODO: Enable characteristic notification
- changeMTU()

Apologies if this is a n00b question but would appreciate any guidance.

**Comment by Reedyuk** [comment]
What exactly are you trying to do? 
sounds like you want to find a device, connect to it and then read a characteristic?

**Reply by carillem** [reply]
Yes, that's correct. So I got up to connecting to the device, finding the services, setting up the characteristics for read (+notification) and write.

**Reply by Reedyuk** [reply]
so what do you want to do next?

**Reply by carillem** [reply]
Negotiate an MTU size and start sending commands to the write characteristic. Just not sure if I'm supposed to be calling any of the commands above first to complete the setup of characteristics and descriptor.

**Reply by Reedyuk** [reply]
best practice is to follow how you would do this on android, generally speaking it should be the same process.
Once you have connected to your peripheral, you can request the mtu size and then you can start to write characteristics.

**Reply by carillem** [reply]
Gotcha. Then, if I want to set notification on my read characteristic, do I then call:
```
notifyCharacteristic(peripheral, readCharacteristic, true)
```

what event does it send back from read? or do I use `readCharacteristic`?

**Reply by Reedyuk** [reply, marked as answer]
Yeah you can use read characteristic or notify characteristic

**Reply by carillem** [reply]
Does `notifyCharacteristic` use the `didCharacteristcValueChanged`?

**Reply by Reedyuk** [reply]
it depends on the platform but i beleive it should do

**Reply by carillem** [reply]
I will give that a try; thank you!

**Reply by carillem** [reply]
Is `writeDescriptor` available in the library?

Nvm. I looked at the code and it's included in `setCharacteristicNotification`.

**Comment by carillem** [comment]
On another note, @Reedyuk, I am looking at the MTU negotiation and it looks like it's not implemented. If I fork and create a PR to add `mtuSize` as a new variable for a `BluetoothPeripheral`, would that align with your design?

**Reply by Reedyuk** [reply]
Yes sounds good to me, you need to make sure it is implemented on at least iOS and Android though.

**Reply by carillem** [reply]
PR is open: https://github.com/Reedyuk/blue-falcon/pull/140

**Reply by Reedyuk** [reply]
thanks, left one comment but its an easy change.
